File: cheat_sheet/python/1_rewrite_brownfield/internal/domain/organization/bot.py
"""Bot domain models."""

from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel


class BotType(StrEnum):
    """Bot type enumeration."""

    LINE = "LINE"
    FACEBOOK = "FB"
    INSTAGRAM = "IG"


class Bot(BaseModel):
    """Bot domain model.

    Represents a channel entity associated with an organization.
    It contains the configuration and credentials for integration across multiple platforms.
    """

    id: int
    organization_id: int
    name: str
    type: BotType
    channel_id: str
    channel_secret: str
    access_token: str
    token_expired_time: datetime
    created_at: datetime
    updated_at: datetime
    expired_at: datetime | None = None
    enable: bool

    def is_active(self) -> bool:
        """Check if the bot is enabled and not expired."""
        if not self.enable:
            return False
        if self.expired_at is not None and datetime.now() > self.expired_at:
            return False
        return True

    def is_token_expired(self) -> bool:
        """Check if the bot's access token has expired."""
        return self.token_expired_time < datetime.now()

    def is_token_valid(self) -> bool:
        """Check if the bot has a valid access token."""
        return self.access_token != "" and not self.is_token_expired()

    class Config:
        """Pydantic configuration."""

        use_enum_values = True

File: cheat_sheet/go/1_rewrite_brownfield/internal/domain/organization/bot.go
package organization

import (
	"time"
)

// Bot represents a channel entity associated with an organization.
// It contains the configuration and credentials for integration across multiple platforms (LINE, Facebook, Instagram).
type Bot struct {
	ID               int        `json:"id"`
	OrganizationID   int        `json:"organization_id"`
	Name             string     `json:"name"`
	Type             BotType    `json:"type"`
	ChannelID        string     `json:"channel_id"`
	ChannelSecret    string     `json:"channel_secret"`
	AccessToken      string     `json:"access_token"`
	TokenExpiredTime time.Time  `json:"token_expired_time"`
	CreatedAt        time.Time  `json:"created_at"`
	UpdatedAt        time.Time  `json:"updated_at"`
	ExpiredAt        *time.Time `json:"expired_at,omitempty"`
	Enable           bool       `json:"enable"`
}

// BotType represents the type of bot.
type BotType string

const (
	BotTypeLine      BotType = "LINE"
	BotTypeFacebook  BotType = "FB"
	BotTypeInstagram BotType = "IG"
	// Add other bot types as needed
)

// IsActive returns true if the bot is enabled and not expired.
func (b *Bot) IsActive() bool {
	if !b.Enable {
		return false
	}
	if b.ExpiredAt != nil && time.Now().After(*b.ExpiredAt) {
		return false
	}
	return true
}

// IsTokenExpired returns true if the bot's access token has expired.
func (b *Bot) IsTokenExpired() bool {
	return b.TokenExpiredTime.Before(time.Now())
}

// IsTokenValid returns true if the bot has a valid access token.
func (b *Bot) IsTokenValid() bool {
	return b.AccessToken != "" && !b.IsTokenExpired()
}
